### ISSUE TITLE
[BUG] Executors on YARN keep getting lost because of error `ERROR NativeDepsLoader: Could not load cudf jni library...`

### ISSUE THREAD

**Issue by Niharikadutta** [NVIDIA/spark-rapids#2163]
**Describe the bug**
I am pretty sure this is something missing with my setup, but I am unable to figure out what and would appreciate any help. I am trying to get Spark to run on YARN using Spark Rapids and I see the executors keep dying with the following error

```
21/04/16 12:38:24 INFO RapidsExecutorPlugin: Initializing memory from Executor Plugin
21/04/16 12:38:45 ERROR NativeDepsLoader: Could not load cudf jni library...
java.io.IOException: Error loading dependencies
	at ai.rapids.cudf.NativeDepsLoader.loadNativeDeps(NativeDepsLoader.java:182)
	at ai.rapids.cudf.NativeDepsLoader.loadNativeDeps(NativeDepsLoader.java:88)
	at ai.rapids.cudf.Cuda.(Cuda.java:24)
	at com.nvidia.spark.rapids.GpuDeviceManager$.setGpuDeviceAndAcquire(GpuDeviceManager.scala:90)
	at com.nvidia.spark.rapids.GpuDeviceManager$.$anonfun$initializeGpu$1(GpuDeviceManager.scala:117)
	at scala.runtime.java8.JFunction1$mcII$sp.apply(JFunction1$mcII$sp.java:23)
	at scala.Option.map(Option.scala:230)
	at com.nvidia.spark.rapids.GpuDeviceManager$.initializeGpu(GpuDeviceManager.scala:117)
	at com.nvidia.spark.rapids.GpuDeviceManager$.initializeGpuAndMemory(GpuDeviceManager.scala:125)
	at com.nvidia.spark.rapids.RapidsExecutorPlugin.init(Plugin.scala:145)
	at org.apache.spark.internal.plugin.ExecutorPluginContainer.$anonfun$executorPlugins$1(PluginContainer.scala:111)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:245)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:245)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:242)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108)
	at org.apache.spark.internal.plugin.ExecutorPluginContainer.(PluginContainer.scala:99)
	at org.apache.spark.internal.plugin.PluginContainer$.apply(PluginContainer.scala:164)
	at org.apache.spark.internal.plugin.PluginContainer$.apply(PluginContainer.scala:152)
	at org.apache.spark.executor.Executor.$anonfun$plugins$1(Executor.scala:234)
	at org.apache.spark.util.Utils$.withContextClassLoader(Utils.scala:221)
	at org.apache.spark.executor.Executor.(Executor.scala:234)
	at org.apache.spark.executor.CoarseGrainedExecutorBackend$$anonfun$receive$1.applyOrElse(CoarseGrainedExecutorBackend.scala:152)
	at org.apache.spark.rpc.netty.Inbox.$anonfun$process$1(Inbox.scala:115)
	at org.apache.spark.rpc.netty.Inbox.safelyCall(Inbox.scala:213)
	at org.apache.spark.rpc.netty.Inbox.process(Inbox.scala:100)
	at org.apache.spark.rpc.netty.MessageLoop.org$apache$spark$rpc$netty$MessageLoop$$receiveLoop(MessageLoop.scala:75)
	at org.apache.spark.rpc.netty.MessageLoop$$anon$1.run(MessageLoop.scala:41)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.concurrent.ExecutionException: java.lang.UnsatisfiedLinkError: /mnt/var/hadoop/tmp/nm-local-dir/usercache/trusted-service-user/appcache/application_1618576629800_0001/container_1618576629800_0001_01_000002/tmp/cudfjni1861444512765804143.so: libnvToolsExt.so.1: cannot open shared object file: No such file or directory
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at ai.rapids.cudf.NativeDepsLoader.loadNativeDeps(NativeDepsLoader.java:180)
	... 34 more
Caused by: java.lang.UnsatisfiedLinkError: /mnt/var/hadoop/tmp/nm-local-dir/usercache/trusted-service-user/appcache/application_1618576629800_0001/container_1618576629800_0001_01_000002/tmp/cudfjni1861444512765804143.so: libnvToolsExt.so.1: cannot open shared object file: No such file or directory
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1934)
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1817)
	at java.lang.Runtime.load0(Runtime.java:810)
	at java.lang.System.load(System.java:1088)
	at ai.rapids.cudf.NativeDepsLoader.loadDep(NativeDepsLoader.java:197)
	at ai.rapids.cudf.NativeDepsLoader.loadDep(NativeDepsLoader.java:211)
	at ai.rapids.cudf.NativeDepsLoader.lambda$loadNativeDeps$1(NativeDepsLoader.java:174)
	... 5 more
21/04/16 12:38:45 ERROR RapidsExecutorPlugin: Exception in the executor plugin
java.lang.UnsatisfiedLinkError: ai.rapids.cudf.Cuda.isPtdsEnabled()Z
	at ai.rapids.cudf.Cuda.isPtdsEnabled(Native Method)
	at ai.rapids.cudf.Cuda.(Cuda.java:32)
```

**Expected behavior**
Ability to run a Spark job successfully using RAPIDS acceleration on a GPU based YARN cluster.

**Environment details**

CUDA version: 11.0
Spark version: 3.0.2
Ubuntu version: 18.04
Resource manager: YARN

Spark configs added to `spark-defaults.conf`:

```
spark.plugins com.nvidia.spark.SQLPlugin
spark.rapids.sql.enabled true
spark.rapids.shims-provider-override com.nvidia.spark.rapids.shims.spark302.SparkShimServiceProvider
spark.rapids.sql.concurrentGpuTasks 2
spark.task.cpus 1
spark.executor.resource.gpu.amount 1
spark.executor.resource.gpu.discoveryScript /opt/spark/examples/src/main/scripts/getGpusResources.sh
spark.jars /opt/spark/jars/rapids-4-spark_2.12-0.4.1.jar,/opt/spark/jars/cudf-0.18.1-cuda11.jar
```
The rapids and cudf jars are present in the `/opt/spark/jars` directory for all nodes.

I have also referred to [this article](https://hadoop.apache.org/docs/r3.1.3/hadoop-yarn/hadoop-yarn-site/UsingGpus.html) and added the following properties to my `yarn-site.xml` file:

```
<property>
      <name>yarn.resource-types</name>
      <value>yarn.io/gpu</value>
</property>
<property>
      <name>yarn.scheduler.capacity.resource-calculator</name>
      <value>org.apache.hadoop.yarn.util.resource.DominantResourceCalculator</value>
</property>
<property>
      <name>yarn.nodemanager.resource-plugins</name>
      <value>yarn.io/gpu</value>
</property>
```
What am I missing?

**Additional context**
I have earlier got it to work on Standalone with above configuration (except obviously the YARN properties).

**Comment by abellina** [comment]
```
Caused by: java.lang.UnsatisfiedLinkError: /mnt/var/hadoop/tmp/nm-local-dir/usercache/trusted-service-user/appcache/application_1618576629800_0001/container_1618576629800_0001_01_000002/tmp/cudfjni1861444512765804143.so: libnvToolsExt.so.1: cannot open shared object file: No such file or directory
```

It seems that the YARN container can't find the CUDA libraries in this case. `libnvToolsExt.so.1` is included in the `lib64` directory. This is what I see in my setup:

```
$ ll /usr/local/cuda-11.0/lib64/|grep libnvToolsExt
     0 lrwxrwxrwx 1 root root        18 May  7  2020 libnvToolsExt.so -> libnvToolsExt.so.1
     0 lrwxrwxrwx 1 root root        22 May  7  2020 libnvToolsExt.so.1 -> libnvToolsExt.so.1.0.0
    36 -rw-r--r-- 1 root root     36832 May  7  2020 libnvToolsExt.so.1.0.0
```

Could it be that the container doesn't have CUDA installed, or it is in a non-standard location?

**Comment by Niharikadutta** [comment]
@abellina Thanks for the prompt response! I see in my node that this file (symlink) does actually exist in the standard location `/usr/local/cuda-11.0/lib64`:

```
/usr/local/cuda-11.0/lib64$ file libnvToolsExt.so.1
libnvToolsExt.so.1: symbolic link to libnvToolsExt.so.1.0.0
```

Could this mean something is wrong with the environment variables `PATH` and `CUDA_PATH`?

**Comment by tgravescs** [comment]
is this running in docker container?  If you are running in the container is cuda-11.0 runtime available in there as well?  just use docker to enter the container and check.

What driver version? (run nvidia-smi to see).

Somewhere either the file doesn't exist or yes its not finding it for some reason.  
> libnvToolsExt.so.1: symbolic link to libnvToolsExt.so.1.0.0
Double check to make sure the libnvToolsExt.so.1.0.0 file is also present.

**Comment by abellina** [comment]
In addition to what @tgravescs suggested, could you run: `ldconfig -p | grep libnvTools`? It would help us figure out where the dynamic linker is attempting to go.

**Comment by Niharikadutta** [comment]
@tgravescs This is not a docker container, just a virtual machine and cuda-11.0 runtime is available in every VM. The driver version on each VM is: `450.51.06`
Also I verified that `libnvToolsExt.so.1.0.0` does exist:
```
lrwxrwxrwx 1 root root    18 May  8  2020 libnvToolsExt.so -> libnvToolsExt.so.1
lrwxrwxrwx 1 root root    22 May  8  2020 libnvToolsExt.so.1 -> libnvToolsExt.so.1.0.0
-rw-r--r-- 1 root root 36832 May  8  2020 libnvToolsExt.so.1.0.0
```
Also is this normal? The 2 layers of symbolic links?

@abellina here is the result of the command `ldconfig -p | grep libnvTools`:
```
$ ldconfig -p | grep libnvTools
        libnvToolsExt.so.1 (libc6,x86-64) => /usr/local/cuda-11.0/targets/x86_64-linux/lib/libnvToolsExt.so.1
        libnvToolsExt.so (libc6,x86-64) => /usr/local/cuda-11.0/targets/x86_64-linux/lib/libnvToolsExt.so
```

**Comment by abellina** [comment]
> Also is this normal? The 2 layers of symbolic links?

Yes that seems fine to me.

The `ldconfig` output seems to indicate that the library is available. One more thing to try to find out what's going on is to run with LD_DEBUG: `spark.executorEnv.LD_DEBUG=all`. This will be verbose, but it will tell you what paths the linker is trying. 

Additionally, unpacking the cudf jar (`jar xvf [cudf jar]`) would give you access to the shared library in `amd64/Linux/libcudfjni.so`. Running `ldd amd64/Linux/libcudfjni.so` would tell you more.

The other possibility here is that executor run-time is different (somehow) than your direct access to the VM. The `LD_DEBUG` way would tell you that via the logs => it should do something unexpected.

**Comment by Niharikadutta** [comment]
Thanks @abellina I'll see about submitting a job via `LD_DEBUG`.

**Comment by Niharikadutta** [comment]
@abellina Quick question regarding the LD_DEBUG config, where can I specify this config? Can I add it to my spark-defaults.conf file like this:
```
spark.executorEnv.LD_DEBUG  all
```

**Comment by abellina** [comment]
Yes, or `--conf spark.executorEnv.LD_DEBUG=all` when you submit the job. Either way it should work.

Sorry for the delay in replying.

**Comment by Niharikadutta** [comment]
I figured out the issue, I am doing a custom (lean) installation of CUDA which evidently did not create the symbolic link to `/usr/local/cuda-11.0`. Once I created that link it started working. Thanks @abellina and @tgravescs for your help! I will close this issue now.

**Comment by abellina** [comment]
Ok thanks @Niharikadutta. Another option may be to run `ldconfig`. It should do the right thing (did for me locally), rather than to have to worry about the symlink to the specific version of cuda.

**Comment by NegatioN** [comment]
I know this might be the wrong place to ask this, however:
if I run into a similar problem, where I'm unable to make a symbolic link, probably because my CUDA installation is in a non-standard place (`/usr/local/nvidia` using Google's standard cuda installer sidecar for Kubernetes containers), is there any way to actually instruct `spark-rapids` to look in that location in the first place? Instead of worrying about `ldconfig` or symlinks?

**Comment by Niharikadutta** [comment]
Just slipping in a question here since my environment is the same as above except I am upgrading Spark version to 3.1.2 which was recently released. Are there any configuration changes I need to make for the new Spark version? I have bumped up the Spark RAPIDS and CuDF versions:

Spark RAPIDS - 0.5.0
CuDF - 0.19.2
CUDA - 11.0

With every other setting being the same, I am seeing the following error on submitting a job:
```
WARN RSCClient: Client RPC channel closed unexpectedly.
```

**Comment by tgravescs** [comment]
Apache spark 3.1.2 is not supported in our released 0.5.0 release because it was not released yet.  It will be officially supported in our next 21.06 release.

**Comment by abellina** [comment]
Adding to the comment from @tgravescs:

You could build 21.06 from source if you want to test early, it should be pretty straightforward `mvn package -DskipTests`.

Just FYI, the driver/executor logs should have more information on what the error actually is. The log message seems to refer to an RPC channel that looks to be maintained by Livy, and likely happens after the real error.

**Comment by Niharikadutta** [comment]
Thanks @tgravescs and @abellina. I saw the Spark312 shim provider override in Rapids 0.5.0 so thought we did some proactive supporting there. What is the tentative release date for 21.06?

**Comment by jlowe** [comment]
> I saw the Spark312 shim provider override in Rapids 0.5.0 so thought we did some proactive supporting there.

We do proactively test with the snapshot versions of Spark as they are being developed so we can stay on top of the latest changes.  We were testing 0.5 against Spark 3.1.2-SNAPSHOT, hence the spark312 shim you saw in the source.

However we cannot release shims that are against Spark snapshot versions because we cannot guarantee that the Spark version doesn't change in a breaking way after our release.  Therefore we explicitly remove snapshot shim versions from the RAPIDS Accelerator jar when it is released.  Snapshot shims are only available during testing before an official release.  Since Spark 3.1.2 had not been officially released when RAPIDS Accelerator 0.5 was released, we did not ship the spark312 shim in 0.5.
